pipeline/lib/pipeline.ml

```diff
@@ -116,23 +126,37 @@ let pipeline ~slack_path ~conninfo ?branch ?pull_number ~dockerfile ~tmpfs
     let+ build_job_id = Current_util.get_job_id current_image
     and+ run_job_id = Current_util.get_job_id current_output
     and+ output = current_output in
+    let output_json_lst = output |> Json_util.parse_many in
+    let benchmark_tbl =
+      let tbl = Hashtbl.create 1000 in
+      List.iteri
+        (fun index output_json ->
+          Hashtbl.add tbl
+            (Yojson.Safe.Util.(member "name" output_json)
+            |> Yojson.Safe.Util.to_string_option)
+            index)
+        output_json_lst;
+      tbl
+    in
     let duration = Ptime.diff (Ptime_clock.now ()) run_at in
     Logs.debug (fun log -> log "Benchmark output:\n%s" output);
     let () =
       let db = new Postgresql.connection ~conninfo () in
-      output
-      |> Json_util.parse_many
+      output_json_lst
       |> List.iter (fun output_json ->
              let benchmark_name =
                Yojson.Safe.Util.(member "name" output_json)
                |> Yojson.Safe.Util.to_string_option
              in
-             Yojson.Safe.Util.(member "results" output_json)
-             |> Yojson.Safe.Util.to_list
-             |> List.map
-                  (Benchmark.make ~duration ~run_at ~repo_id ~benchmark_name
-                     ~commit ?pull_number ?build_job_id ?run_job_id ?branch)
-             |> List.iter (Models.Benchmark.Db.insert db));
+             match is_duplicate benchmark_name benchmark_tbl with
+             | Error e -> raise @@ Failure e
```

Comment:
The `raise` can be done while building the Hashtbl (we can detect that there is a duplicate before inserting)